django/db/models/query.py

```diff
@@ -942,6 +942,8 @@ def extra(self, select=None, where=None, params=None, tables=None,
 
     def reverse(self):
         """Reverse the ordering of the QuerySet."""
+        if not self.query.can_filter():
+            raise TypeError('Cannot reverse a query once a slice has been taken')
```

Comment:
I'm not sure if using assert is best since that's ignored when using `python -O`.